src/circuit/v2/util.ts

```diff
@@ -0,0 +1,87 @@
+import { abortableDuplex, AbortError } from 'abortable-iterator'
+import { pipe } from 'it-pipe'
+import type { Duplex, Sink, Source } from 'it-stream-types'
+import { TimeoutController } from 'timeout-abort-controller'
+import { Uint8ArrayList } from 'uint8arraylist'
+import type { Limit } from './pb/index.js'
+import { logger } from '@libp2p/logger'
+
+const log = logger('libp2p:circuit:v2:util')
+
+type DuplexStream = Duplex<Uint8ArrayList, Uint8ArrayList | Uint8Array>
+export async function createLimitedShortCircuit (source: DuplexStream, destination: DuplexStream, limit?: Limit) {
+  // trivial case
+  if (limit == null) {
+    return await pipe(source, destination, source)
+  }
+  const src = durationLimitDuplex(dataLimitDuplex(source, Number(limit.data ?? 0)), limit.duration ?? 0)
+  const dst = durationLimitDuplex(dataLimitDuplex(destination, Number(limit.data ?? 0)), limit.duration ?? 0)
```

Comment:
We shouldn't downcast from `BigInt` to `Number`, otherwise we can't satisfy the limits that may be set.